### ISSUE TITLE
@skip not working like it used to moving from Thucydides to Serenity 

### ISSUE THREAD

**Issue by joecolantonio** [serenity-bdd/serenity-core#364]
Hi,

I finally convinced my team to move over from Thucydides :

   <thucydides.version>0.9.229</thucydides.version>
   <thucydides.jbehave.version>0.9.229</thucydides.jbehave.version>

To

   <serenity.version>1.1.21</serenity.version>
   <serenity.jbehave.version>1.5.0</serenity.jbehave.version>

I also tried with 

 <serenity.version>1.1.29-rv.1</serenity.version>
 <serenity.jbehave.version>1.7.0</serenity.jbehave.version>

But teams are now complaining that previous feature files that had Meta:@skip are now running steps when it did not before serenity. For example

In Thucydides this:

@MANUAL_11111
Scenario: Help document content verification
Meta:
@skip
Given user is using application
When a user selects help
Then Verify help document

Would skip all the G/W/T even if there was an existing step definitions that match on one of the G/W/T

Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 16.935 sec

NOW in Serenity it will actually run the Given and the When step and show that the test skipped even though it did run:

TEST STARTED: Help document content verification
--------------------------------------------------------------------
Meta:
@skip 

[pool-1-thread-1] INFO net.serenitybdd.core.Serenity - TEST IGNORED
Given user is using application
 INFO [NotificationPage:63] 2016-03-18 14:07:10,905 - clickOnViewerDisabledModal(): No modal dialog to close
 INFO [MastheadPage:162] 2016-03-18 14:07:12,275 - CLICK ON USERNAME : START
 INFO [MastheadPage:171] 2016-03-18 14:07:12,486 - CLICK ON USERNAME : END
When a user selects help
Then Verify help document (PENDING)
[pool-1-thread-1] INFO net.serenitybdd.core.Serenity - TEST PENDING
@Then("Verify help document")
@Pending
public void thenVerifyHelpDocument() {
  // PENDING
}

Any ideas how to fix before I have to revert everything back to Thucydides again?

**Comment by YamStranger** [comment]
Hello @joecolantonio 

Thanks for this issue. 

I checked it with  https://github.com/serenity-bdd/serenity-test-projects/tree/master/web-todomvc-tests
and 
1.1.29-rv.1
1.7.0

executed test only for one module: 
```
./gradlew clean :jbehave-without-examples:test aggregate
``` 

I see that serenity-jbehave mark these tests as ignored, but during initialization open browser and immediately close it without performing any action. 

Did your tests execute fully or only open browser and close it? 

Thanks!

**Comment by YamStranger** [comment]
@joecolantonio 
Attached output with marked tests as pending 
[out.txt](https://github.com/serenity-bdd/serenity-core/files/180449/out.txt)

Sure it will be fixed soon - problem with initialization. 

Just want to be sure that it is only one problem.

**Comment by joecolantonio** [comment]
@YamStranger - yes it opens the browser _and_ performs a login and navigates and enters data for the Given and When steps because they have step definitions that other non skip scnearios use. But since this is marked as @skip it should not run these steps. I did not have this issue with the same .feature files with Thucydides. In Thucydides it completely skipped these scenarios.

**Comment by YamStranger** [comment]
@joecolantonio Thanks for this answer. 
I can not reproduce your issue for now, please give me some additional details: 
1) `.feature` files - so you use cucumber in the under the hood, not jbehave? 
2) what build system do you use, and is it possible add here your build configuration (pom or build.gradle file) - if not please attach dependencies tree. 

Thanks!

**Comment by joecolantonio** [comment]
@YamStranger I will send more info on Monday when I'm in the office. Yes we use jbehave NOT cucumber. Because I work for a large company our standard is to use. feature rather than .stories because we have other teams that do use cucumber/specflow. So we do use jbehave and it does work with our story files with the .feature extension. We do use Maven I will send pom.file on Monday as well

**Comment by YamStranger** [comment]
Thanks @joecolantonio!!

Also please add serenity bdd configurations if you use that (properties, conf, anything that used for configuration serenity BDD). And It will be great if you give me some tips how you make serenity-jbehave to work with feature files (by default only .story files are used, and I can not find any direct way to change this behaving)

A am sure that this issue will be fixed fast, after I find way to reproduce it. 

Thanks!

**Comment by joecolantonio** [comment]
Hi @YamStranger 

I've attached the projects pom.xml as well as a .feature file example along with its output. To run I use the following:

clean verify serenity:aggregate -U -Dmaven.test.failure.ignore=true -Dwebdriver.driver=chrome -Dwebdriver.chrome.driver=./chromedriver.exe -Dwebdriver.base.url=https://1.111.111.111:0000 -Dstory.timeout.in.secs=300 -Dignore.failure.in.stories=true

and the serenity.properites files has this:

serenity.take.screenshots=FOR_FAILURES
story.timeout.in.secs=28800
webdriver.base.url=http://127.0.0.1:9000
maven.test.failure.ignore=true

Thanks for looking into this!!!

[TestSuiteRunner.txt](https://github.com/serenity-bdd/serenity-core/files/182289/TestSuiteRunner.txt)

**Comment by YamStranger** [comment]
@joecolantonio Thanks!

**Comment by YamStranger** [comment]
Hello @joecolantonio

As I understand you configure paths to jbehave sotries (features in your case) manually somehow from file `testsuite.properties`.

It seems that serenity-jbehave was updated, so to run your tests with your own paths of stories you should update RadDesktopTestSuite. After it your tests with @skip will be skipped immediately, without any interactions described in your tests.
We still have minor exception (serenity-bdd/serenity-jbehave/issues/63): @Managed driver will be initialized even in @skip scenario, but closed immediately without any additional actions - it will take almost zero seconds (please find details below).  

**Shortly**: 
```
public class RadDesktopTestSuite extends SerenityStories {
    @Override
    public List<String> storyPaths() {
        return Arrays.asList(
                "stories/todo_by_examples/one_user_add_todo_using_examples.example",
                "stories/todo_multiple_operations/one_user_add_todo_many_times.example",
                "stories/todo_single_operations/one_user_add_todo_once.example"
        );
    }
}
```
where `List<String> storyPaths` should return file paths to each story that should be executed. 
And please updatedJunit from 4.11 to 4.12

### Detailed steps to run example of configuration with custom jbehave story path resolution
I will explain all based on  [web-todomvc-tests](https://github.com/serenity-bdd/serenity-test-projects/tree/master/web-todomvc-tests) project to be sure that it can be reproducible. 

#### Preconditions 
First of all I updated in build.gradle versions:
 -   serenity-core with version `1.1.29-rc.1`
 -  serenity-jbehave with version `1.7.0`

initially, after commands:
```
git clone https://github.com/serenity-bdd/serenity-test-projects.git
cd serenity-test-projects/
git submodule update --init --recursive
```
directory `serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/java/net/serenitybdd/demos/todos`
will contains files that should be removed (they are used to find standard story files): 
 - OneUserAddTodoManyTimes
 - OneUserAddTodoOnce
 - OneUserAddTodoUsingExamples

and for next files lets change extension from `.story` to `.example`:
 - serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/resources/stories/todo_by_examples/one_user_add_todo_using_examples.story
- serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/resources/stories/todo_multiple_operations/one_user_add_todo_many_times.story
- serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/resources/stories/todo_single_operations/one_user_add_todo_once.story

after that lets run command `./gradlew clean :jbehave-without-examples:test aggregate` in folder `serenity-test-projects/web-todomvc-tests`. 
Serenity BDD with Jbehave should not find any tests to execute: 
![before1](https://cloud.githubusercontent.com/assets/1667699/13964126/fce45a26-f070-11e5-86ec-7946ae9d3255.png)
![before](https://cloud.githubusercontent.com/assets/1667699/13964127/fce7e2ae-f070-11e5-8f0a-dc4ad0a6a968.png)

#### Updating tests to use custom jbehave story path resolution  

Now we will update our tests to run jbehave stories based on provided paths. After **Preconditions** we have some story files, with custom name and extension (I used `.example` and not `.feature` because of cucumber in class path, such file extension will broke report generation) .

Lets add to directory `serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/java/net/serenitybdd/demos/todos` class to provide paths to our custom stories: 
```
public class RadDesktopTestSuite extends SerenityStories {
    @Override
    public List<String> storyPaths() {
        return Arrays.asList(
                "stories/todo_by_examples/one_user_add_todo_using_examples.example",
                "stories/todo_multiple_operations/one_user_add_todo_many_times.example",
                "stories/todo_single_operations/one_user_add_todo_once.example"
        );
    }
}
```
also lest update one story at path `serenity-test-projects/web-todomvc-tests/jbehave-without-examples/src/test/resources/stories/todo_by_examples/one_user_add_todo_using_examples.example` to make sure that @skip work correctly: 

```
Feature: Help -  Ability to retrieve the help document

Narrative:
In order check reports
As an Developer
I want to workin with todos using examples

@MANUAL_111
Scenario: Record a new todo action for future use using examples
Meta:
@skip

Given I need to <definition>
When I add the todo action "<action>"
Then "<action>" should be recorded in the "Active" items
Examples:
|definition|action|
|buy some milk|Buy some milk|
|buy some sugar|Buy some sugar|
|buy some tea|Buy some tea|
```
and execute `./gradlew clean :jbehave-without-examples:test aggregate` in folder `serenity-test-projects/web-todomvc-tests`.
Here is report:
![after2](https://cloud.githubusercontent.com/assets/1667699/13964478/8f0b85fe-f072-11e5-825c-7ee55bc6ae22.png)
![after](https://cloud.githubusercontent.com/assets/1667699/13964477/8f085032-f072-11e5-8537-acd23bd6d13a.png)

as you see serenity-jbehave found stories, executed all of them and skipped one scenario.
_____________
Lets mark all scenarios with @skip:

![after](https://cloud.githubusercontent.com/assets/1667699/13964679/995042a6-f073-11e5-84d7-f6f1d70f5430.png)
![after1](https://cloud.githubusercontent.com/assets/1667699/13964680/9951509c-f073-11e5-990c-794c4a06afb5.png)
All scenarios was skipped without execution of all steps. 

_____________
More that that, lets remove @skip tag: 
```
Feature: Help -  Ability to retrieve the help document

Narrative:
In order check reports
As an Developer
I want to workin with todos using examples

@MANUAL_111
Scenario: Record a new todo action for future use using examples

Given I need to <definition>
When I add the todo action "<action>"
Then "<action>" should be recorded in the "Active" items
Examples:
|definition|action|
|buy some milk|Buy some milk|
|buy some sugar|Buy some sugar|
|buy some tea|Buy some tea|
```  

now serenity-bdd executed all tests:
![after3](https://cloud.githubusercontent.com/assets/1667699/13964581/15e3a534-f073-11e5-96eb-c8df81a87fa2.png)

**Comment by joecolantonio** [comment]
@YamStranger thanks for the detailed workaround! Unfortunately for some reason the behavior is still occurring for me. 
+ I updated my pom with 1.1.29-rv.1 & 1.7.0
+ I replaced the code I had in my RadDesktopTestSuite with the code you send and replaced the path with mine:

public class RadDesktopTestSuite extends SerenityStories {
    @Override
    public List<String> storyPaths() {
        return Arrays.asList(
                "jbehave/stories/ui/masthead/Help.feature"
       
        );
    }
}

I ran with my maven command:

clean verify serenity:aggregate -U -Dmaven.test.failure.ignore=true -Dwebdriver.driver=chrome -Dwebdriver.chrome.driver=./chromedriver.exe -Dwebdriver.base.url=https://1.111.111.111:0000 -Dstory.timeout.in.secs=300 -Dignore.failure.in.stories=true

The @skip steps still executes the Given and When step definitions and it logins in and select help.

When i look at the report it shows that it SKIPPED but I know that it ran them 

The only workaround I could come  up with is when I run from maven to pass it _-Dmetafilter=-skip_ but dont think that is an optimal solution. I could make a quick video of my behavior is that would help?

Thanks again for all your time!

**Comment by YamStranger** [comment]
@joecolantonio no, video will not help ))

Will try to create example project with your pom and  feature, and post it on github. 

Hope it will help...

**Comment by YamStranger** [comment]
@joecolantonio 

This bug can be reproduced only if @Managed webdrivers are used with  gradle and mvn. 

If in tests used PageObjects (as in described example above) all work fine. 

Thanks. Will updated this ticket as soon as it will be fixed.

**Comment by YamStranger** [comment]
@joecolantonio this bug will be fixed using filtering feature of jbehave, so behaving will be the same as your workaround with `-Dmetafilter=-skip `. (same for @ignored). 

Difference will be only in report - if you will use these workaround nothing about those scenarios/stories will be in serenity bdd report. 

After fix such scenarios/stories will be included in report, and marked as skipped or ignored as well. 

Thanks!

**Comment by joecolantonio** [comment]
Awesome @YamStranger thanks again! I need to see how I can contribute to this project since I really do love Serenity. Thanks again

**Comment by YamStranger** [comment]
Issue fix will available in next serenity-jbehave release.

**Comment by YamStranger** [comment]
@joecolantonio please check fix of this issue in latest jbehave (1.9.0) and serenity core (1.1.29-rc.3)

**Comment by YamStranger** [comment]
It seems that this issue already fixed and ticket will be closed.
